src/client/common/utils/version.ts

```diff
@@ -3,7 +3,290 @@
 
 'use strict';
 
+// tslint:disable:no-multiline-string
+
 import * as semver from 'semver';
+import { verboseRegExp } from './regexp';
+
+//===========================
+// basic version info
+
+/**
+ * basic version information
+ *
+ * A normalized object will only have non-negative numbers, or `-1`,
+ * in its properties.  A `-1` value is an indicator that the property
+ * is not set.  Lower properties will not be set if a higher property
+ * is not.
+ *
+ * Note that any object can be forced to look like a VersionInfo and
+ * any of the properties may be forced to hold a non-number value.
+ * To resolve this situation, pass the object through
+ * `normalizeVersionInfo()` and then `validateVersionInfo()`.
+ */
+export type BasicVersionInfo = {
+    major: number;
+    minor: number;
+    micro: number;
+    // There is also a hidden `unnormalized` property.
+};
+
+function normalizeVersionPart(part: number): number {
```

Comment:
```suggestion
function normalizeVersionPart(part: number | string | undefined | null): number {
```

with the way it is right now the compiler is doing weird stuff, and you have to do `(part as unknown) as string`:

<img width="461" alt="image" src="https://user-images.githubusercontent.com/51720070/91365910-1a083b80-e7b7-11ea-9194-b62a2ddcf099.png">

Reply:
We are expecting it to be a number (from `VersionInfo.major`, etc.).  However, some potential sources of the raw data may work around type-checking, so I've taken a defensive approach during normalization.  Consequently, we are ensuring that the properties of the normalized object not only have valid values, but that the values are of the correct type.

Reply:
Thinking about it for a second more 😄, I think I agree with your recommendation.  For exported helpers, though, we should preserve the stricter type expectation.